### PR TITLE
Covid Self Certify initial commit

### PR DESCRIPTION
**Pull request by Ajcitrite**

### CODE REVIEW

File: bundles/http/Citrix/Covid-19-Self-Certify/script.js [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+                { name: "appId", type: "STRING" }
+            ],
+            function: covidSelfCertify
+        }
+    ]
+})
+
+async function validateResponse(response){
+    let responseBody, errorMessage, errorStatus
+    if(response.headers.map['content-type'].includes("application/json")){
```
Comment (vojtacitrite) [inline bundles/http/Citrix/Covid-19-Self-Certify/script.js:65]: case sensitive operation but headers are case insensitive, header value could have multiple parameters separated by semicolon, example of valid HTTP header: `Content-Type: application/json; charset=utf-8`

File: bundles/http/Citrix/Covid-19-Self-Certify/script.js [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+                { name: "city", type: "STRING", required: true },
+                { name: "appId", type: "STRING" }
+            ],
+            function: covidSelfCertify
+        }
+    ]
+})
+
+async function validateResponse(response){
+    let responseBody, errorMessage, errorStatus
```
Comment (vojtacitrite) [inline bundles/http/Citrix/Covid-19-Self-Certify/script.js:64]: declaration without initialization, it's Javascript legacy approach for `var` which is not needed when `let` and `const` are available. could be easily refactored - see following comments

File: bundles/http/Citrix/Covid-19-Self-Certify/script.js [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+    if(response.headers.map['content-type'].includes("application/json")){
+        const jsonBody = await response.json()
+        if(!response.ok){
+            responseBody = jsonBody
```
Comment (vojtacitrite) [inline bundles/http/Citrix/Covid-19-Self-Certify/script.js:68]: return ({ responseBody: jsonBody, errorMessage: jsonBody.error_description, errorStatus: true})

File: bundles/http/Citrix/Covid-19-Self-Certify/script.js [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+            errorMessage = jsonBody.error_description
+            errorStatus = true
+        }else{
+            responseBody = jsonBody
```
Comment (vojtacitrite) [inline bundles/http/Citrix/Covid-19-Self-Certify/script.js:72]: return ({ responseBody: jsonBody, errorStatus: false})

File: bundles/http/Citrix/Covid-19-Self-Certify/script.js [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+            errorStatus = false
+        }
+    }else{
+        const body = await response.text()
+        responseBody = body
+        errorMessage = 'Unable to proccess the request'
+        errorStatus = true
+    }
+    return {responseBody,errorMessage,errorStatus}
```
Comment (vojtacitrite) [inline bundles/http/Citrix/Covid-19-Self-Certify/script.js:81]: remove this `return`

File: bundles/http/Citrix/Covid-19-Self-Certify/script.js [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+        if(!response.ok){
+            responseBody = jsonBody
+            errorMessage = jsonBody.error_description
+            errorStatus = true
+        }else{
+            responseBody = jsonBody
+            errorStatus = false
+        }
+    }else{
+        const body = await response.text()
```
Comment (vojtacitrite) [inline bundles/http/Citrix/Covid-19-Self-Certify/script.js:76]: bug, response body is a buffer which is already processed and will throw an error when accessed again.

File: bundles/http/Citrix/Covid-19-Self-Certify/script.js [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+                limit: limit,
+                offset: offset,
+                sort_by: "app_item_id",
+                sort_desc: true
+            })
+        })
+        const {responseBody,errorMessage,errorStatus} = await validateResponse(itemsResponse)
+        if (errorStatus) {
+            console.log(JSON.stringify(responseBody),errorMessage)
+            throw new Error(JSON.stringify(errorMessage))
```
Comment (vojtacitrite) [inline bundles/http/Citrix/Covid-19-Self-Certify/script.js:109]: no need to `JSON.stringify`

**Review by vojtacitrite — CHANGES_REQUESTED** [review]

File: bundles/http/Citrix/Covid-19-Self-Certify/script.js (hunk 1 above)
Comment (Ajcitrite) [reply bundles/http/Citrix/Covid-19-Self-Certify/script.js:65]: For case-sensitive problem, We have converted the header's content type to LowerCase, 
If there are multiple content-type we only need to check that 'application/json' is in the string or not.

File: bundles/http/Citrix/Covid-19-Self-Certify/script.js (hunk 7 above)
Comment (Ajcitrite) [reply bundles/http/Citrix/Covid-19-Self-Certify/script.js:102]: Changes Done

File: bundles/http/Citrix/Covid-19-Self-Certify/script.js (hunk 6 above)
Comment (Ajcitrite) [reply bundles/http/Citrix/Covid-19-Self-Certify/script.js]: Changes Done

File: bundles/http/Citrix/Covid-19-Self-Certify/script.js (hunk 5 above)
Comment (Ajcitrite) [reply bundles/http/Citrix/Covid-19-Self-Certify/script.js]: Changes Done

File: bundles/http/Citrix/Covid-19-Self-Certify/script.js (hunk 4 above)
Comment (Ajcitrite) [reply bundles/http/Citrix/Covid-19-Self-Certify/script.js]: Changes Done

File: bundles/http/Citrix/Covid-19-Self-Certify/script.js (hunk 3 above)
Comment (Ajcitrite) [reply bundles/http/Citrix/Covid-19-Self-Certify/script.js]: Changes Done ✅

File: bundles/http/Citrix/Covid-19-Self-Certify/script.js (hunk 2 above)
Comment (Ajcitrite) [reply bundles/http/Citrix/Covid-19-Self-Certify/script.js]: Changes Done ✅

File: bundles/http/Citrix/Covid-19-Self-Certify/script.js [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+                sort_desc: true
+            })
+        })
+        const {responseBody,errorMessage,errorStatus} = await validateResponse(itemsResponse)
+        if (errorStatus) {
+            console.log(JSON.stringify(responseBody),errorMessage)
+            throw new Error(errorMessage)
+        }
+        total = responseBody.total
+        const items = responseBody.items ?? []
```
Comment (vojtacitrite) [inline bundles/http/Citrix/Covid-19-Self-Certify/script.js:105]:
```
const _ = library.load('lodash')
const items = (responseBody.items || []).map(item => ({
    employeeName: getTextValue(item.fields, 'text', 'Name',
    email:   // etc
})
const uniqueItems = _.uniqueBy(items, 'email')
dataStore.save('items', uniqueItems)
```
+ delete unused `userEmails`

File: bundles/http/Citrix/Covid-19-Self-Certify/script.js [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+async function covidSelfCertify({ client, dataStore, integrationParameters, actionParameters }) {
+    if (actionParameters.appId !== null) {
+        dataStore.deleteById('items', actionParameters.email)
+    }
+    const response = await client.fetch(`item/app/${integrationParameters.covidAppId}`, {
+        method: "POST",
+        body: JSON.stringify({
+            "fields": {
+                "email": [{ "type": "work", "value": actionParameters.email }],
+                "title": [{ "value": actionParameters.email.substring(0, actionParameters.email.indexOf('@')).split('.').reduce((initialvalue, currentvalue) => { return initialvalue + (currentvalue[0].toUpperCase() + currentvalue.substring(1, currentvalue.length) + ' ') }, '').trimEnd() }],
```
Comment (vojtacitrite) [inline bundles/http/Citrix/Covid-19-Self-Certify/script.js:138]: line too long: refactor title extraction to dedicated function

File: bundles/http/Citrix/Covid-19-Self-Certify/script.js [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+            console.log(JSON.stringify(responseBody),errorMessage)
+            throw new Error(errorMessage)
+        }
+        total = responseBody.total
+        const items = responseBody.items ?? []
+        items.forEach(item => {
+            const fields = item.fields ?? []
+            const region = getTextValue(item.fields, 'text', 'Region') ?? null
+            const countries = String(getTextValue(item.fields, 'text', 'Countries')).split(',')
+            countries?.forEach(country => {
```
Comment (vojtacitrite) [inline bundles/http/Citrix/Covid-19-Self-Certify/script.js:177]: if some `item` has the same countries will it create duplicate countries records or not?

File: bundles/http/Citrix/Covid-19-Self-Certify/script.js [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+    return item?.values[0]?.value ?? null
+}
+
+function getCategoryValue(dataItems) {
+    const confirmation = dataItems.find(item => item.type === 'category')
+    return confirmation?.values[0]?.value?.text ?? null
+}
+
+function escapeString(conditions) {
+    return conditions?.replaceAll('<br/>', '\r\n').replaceAll('</p>', '\r\n').replaceAll('<p>', '') ?? null
```
Comment (vojtacitrite) [inline bundles/http/Citrix/Covid-19-Self-Certify/script.js:208]: better function name: this algorithm is some html to plain text but definitely not string escaping

File: bundles/http/Citrix/Covid-19-Self-Certify/script.js [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+}
+
+function getTimestamp(dataItems) {
+    const date = dataItems.find(item => item.type === 'date')
+    return date != undefined ? new Date(date.values[0].start) : null
```
Comment (vojtacitrite) [inline bundles/http/Citrix/Covid-19-Self-Certify/script.js:213]: better function name: `timestamp` is in civil and IT world has differnt meaning a little, `timestamp` usually refers to Unix timestamp but this function returns `Date`

**Review by vojtacitrite — CHANGES_REQUESTED** [review]

File: bundles/http/Citrix/Covid-19-Self-Certify/script.js (hunk 8 above)
Comment (Ajcitrite) [reply bundles/http/Citrix/Covid-19-Self-Certify/script.js:105]: Changes done ✅

File: bundles/http/Citrix/Covid-19-Self-Certify/script.js (hunk 9 above)
Comment (Ajcitrite) [reply bundles/http/Citrix/Covid-19-Self-Certify/script.js:138]: Implemented a new function `getTitleFromEmail`.

File: bundles/http/Citrix/Covid-19-Self-Certify/script.js (hunk 10 above)
Comment (Ajcitrite) [reply bundles/http/Citrix/Covid-19-Self-Certify/script.js:177]: We have used `uniq` function to remove the duplicate countries.

File: bundles/http/Citrix/Covid-19-Self-Certify/script.js (hunk 11 above)
Comment (Ajcitrite) [reply bundles/http/Citrix/Covid-19-Self-Certify/script.js:208]: Function name has been changed to `convertHtmlTagsToStringEscapeSequence`

File: bundles/http/Citrix/Covid-19-Self-Certify/script.js (hunk 12 above)
Comment (Ajcitrite) [reply bundles/http/Citrix/Covid-19-Self-Certify/script.js:213]: We are getting the timestamp from the api, even thought the `item.type` is `date`
For example : `2021-08-06T13:38:57Z`

File: bundles/http/Citrix/Covid-19-Self-Certify/script.js (hunk 10 above)
Comment (vojtacitrite) [reply bundles/http/Citrix/Covid-19-Self-Certify/script.js:184]: it's unique per `item.fields` but not per all `items`, right?

**Review by vojtacitrite — APPROVED** [review]

File: bundles/http/Citrix/Covid-19-Self-Certify/metadata.json [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+  "i18nLanguages" : [ "en" ],
+  "apps" : [ {
+    "uuid" : "e0fc92f2-8671-4725-91e1-05fa60af91be",
+    "trackingUuid" : "deec9060-18e0-4bc9-9b82-3c38af0cc77d",
+    "title" : "Self Certify",
+    "numberOfEvents" : 4,
+    "action" : true
+  } ],
+  "vaResolvers" : [ ],
+  "tags" : [ ],
```
Comment (LubosTomandl) [inline bundles/http/Citrix/Covid-19-Self-Certify/metadata.json:22]: Please add the helpLink for documentation before I can merge. Thanks!

**Review by LubosTomandl — CHANGES_REQUESTED** [review]

**Review by LubosTomandl — APPROVED** [review]

File: bundles/http/Citrix/Covid-19-Self-Certify/metadata.json (hunk 13 above)
Comment (Ajcitrite) [reply bundles/http/Citrix/Covid-19-Self-Certify/metadata.json]: Added the documentation link in metadata.json file.

**Pull request merged**